### ISSUE TITLE
[BUG]: Spyro Enter the Dragonfly cutscene audio is delayed and cut

### ISSUE THREAD

**Issue by CRASHARKI** [PCSX2/pcsx2#8286]
### Describe the Bug

In Spyro Enter the Dragonfly all the cutscenes (Intros, midgame and endings) have the audio delayed and the cutscenes end before the audio does. This might affect other sounds in the game, but it's most noticeable in the cutscenes since they requiere sync. This issue seems to be present since the game was made playable in PCSX2.

### Reproduction Steps

Launch the game and start a new game to watch the intro cutscene with audio desynced and cut.

These are a couple of videos showcasing the issue, they are old, but the issue is still the same.
https://www.youtube.com/watch?v=q2ndlQNMWoc
https://www.youtube.com/watch?v=sdJWeTqsolw

### Expected Behavior

Audio playing at the correct time.

Here's an old video recorded from an actual PS2 showing the audio playing correctly:
https://www.youtube.com/watch?v=-xJCAQMnhhs

### PCSX2 Revision

v1.7.4156

### Operating System

Windows 10 (64bit)

### CPU

i7-6700K

### GPU

NVIDIA GeForce GTX 960

### GS Settings

It happens in both hardware and software.

### Emulation Settings

All settings are default.

**Comment by Buzzardsoul** [comment]
Likely due to the fact that this game is still using a 9 year old patch.

**Comment by CRASHARKI** [comment]
The patch was to prevent a crash/black screen on startup, but I don't know if it was related to audio or file loading.

**Comment by Buzzardsoul** [comment]
@CRASHARKI It's been known that the patch is very janky for a long time but never fixed due to the issue being extremely inconsequential.

**Comment by Mrlinkwii** [comment]
@Buzzardsoul  sorry what

**Comment by Buzzardsoul** [comment]
> @Buzzardsoul sorry what

To put it simply, this issue is entirely caused by the Startup patch that the game still needs in order to work.

It's not a fault of the SPU2 nor the GS emulation.

**Comment by CRASHARKI** [comment]
@RedDevilus But this game has no FMVs, unless you mean the in-game cutscenes.

**Comment by Mrlinkwii** [comment]
@CRASHARKI  The patch fixes the crash that the game encounters when you get to the title screen

**Comment by RedDevilus** [comment]
> @RedDevilus But this game has no FMVs, unless you mean the in-game cutscenes.

Yeah, meant the cutscenes. Force of habit to say FMV even if difference between pre-rendered and real-time, my bad.

**Comment by Buzzardsoul** [comment]
I really don't feel like arguing, therefore I will just post my proof that it's quite literally a GIF issue that's being ignored by the patch itself.

This is the offset the game gets stuck on (PAL)

![untitled](https://user-images.githubusercontent.com/126727903/222520410-673c0911-2305-4339-941e-c562140e9d62.PNG)

**Comment by CRASHARKI** [comment]
Update: I made a few tests with PCSX2 1.6.0 (Since that one allows to disable Automatic Gamefixes, I don't think nightly builds do, and the issue still occurs in 1.6.0), with Automatic Gamefixes off the main menu sound loads at the right time instead of delayed but the screen stays black, as soon as you activate Automatic Gamefixes with the game still on, the screen will show up like normal (Although the sound won't sync obviously). If you now disable them and start a new game, the game will play as normal and the audio will sync just fine with no more black screens. I don't know if this could be automatically applied through patches though, but at leats now we know more about how the patch is messing with the audio. Could there be a way to disable the code in the patch after it's been applied once maybe?

**Comment by Goatman13** [comment]
> I really don't feel like arguing, therefore I will just post my proof that it's quite literally a GIF issue that's being ignored by the patch itself.
> 
> This is the offset the game gets stuck on (PAL)
> 
> ![untitled](https://user-images.githubusercontent.com/126727903/222520410-673c0911-2305-4339-941e-c562140e9d62.PNG)

Game just wait there for Vblank, and continue when irq happen.

> Update: I made a few tests with PCSX2 1.6.0 (Since that one allows to disable Automatic Gamefixes, I don't think nightly builds do, and the issue still occurs in 1.6.0), with Automatic Gamefixes off the main menu sound loads at the right time instead of delayed but the screen stays black, as soon as you activate Automatic Gamefixes with the game still on, the screen will show up like normal (Although the sound won't sync obviously). If you now disable them and start a new game, the game will play as normal and the audio will sync just fine with no more black screens. I don't know if this could be automatically applied through patches though, but at leats now we know more about how the patch is messing with the audio. Could there be a way to disable the code in the patch after it's been applied once maybe?

I can't reproduce your test result. Code patch applied once stays there even if you disable it in menu. Unless game overwrite that memory area, which isn't the case here (I tested this). 

Current patch return true in SpyroWorld::CinematicsMusicStreamReady() [0x1E7070]. Which normally fail because JukeBox::GetMusicStreamStatus() [0x27EA50] return false. JukeBox::GetMusicStreamStatus() return val_from_iop & 0x200. On IOP side there seems to be only one place where that can be set, function IsChannelPlaying [0x8D1BC] should return 1 (under the hood its u16 result = sceSdGetParam(0x502), which in turn just read 0x1F90001A), result is then shifted left by 9 and should end up as 0x200 eventually. This get later ored to current status, and send by sif to EE which expect that mentioned 0x200 to be true. But function IsChannelPlaying return 0x4000 for some reason, which shifted by 9 is 0x800000. So obviously JukeBox::GetMusicStreamStatus() return 0 at this point. Anyway. Forcing return true at any stage of that chain have same effect as current patch. 

Issue could be anything, but it seems that game want that reg to be set to 1. I tried to read SPU2 code to figure out little bit more, but its kinda hard part of pcsx2 code to read. This seems to be reading ADSR.Value for core 0 voice 1. But only place that can set this value to non 0/1 seems to be WriteRegPS1? Maybe there is a way to set it directly from DMA or something?
For people that want to investigate this. 0x921C0 is where game call IsChannelPlaying on IOP side. 
Given memory offsets are for SLUS-20315 version of game.

Sorry for wall of text and little off-topic, since my post explain why patch is there and not why there is desync between a/v.

**Comment by stenzek** [comment]
Reading ADSR volume to know which channels are busy was very common in PS1 games, wouldn't be surprised if the idea carried over. I suppose the question here would be why the voice isn't busy in the first place.

**Comment by CRASHARKI** [comment]
> I can't reproduce your test result. Code patch applied once stays there even if you disable it in menu. Unless game overwrite that memory area, which isn't the case here (I tested this).

I forgot to mention I'm using PAL version, and I can reproduce it very consistently. Although when I activate and deactivate the patch too fast or at certain points in the loading the delay happens too, I don't really know why.

**Comment by Buzzardsoul** [comment]
Ok, after doing some hardware tests on the PAL version, it seems I may have found out why the patch is needed just for curiosity's. The offset "0x2038A290" Is a read/write offset. On a real ps2, the value is "0x007FFE00" But on PCSX2, the value is "0x00800000" and thus, the offset of "0x0027EF80" will read a value that isn't correct.

**Comment by Goatman13** [comment]
With hardware test result from Buzzardsoul, i figured out that "ADSR Value" on real hardware is 0x3FFF, while on pcsx2 it is 0x4000. Because value is left shifted by 9 before sending to EE. 0x4000 << 9 = 800000, but 0x3FFF << 9 give exactly 0x7FFE00. So it's a matter of 1. Remember that game do stupid check and test this value with 0x200 to continue or not.

On pcsx2 i narrowed down issue to be `V_ADSR::Calculate() case2 // decay` issue. Currently when we enter that state in Spyro game, Value is 0x7FFFFFFF, DecayRate is 0, calculated "off" is 0xC, which in turn mean we are going to use PsxRates[0x90] to decrease ADSR Value. PsxRates[0x90] = 0x3FFFFFFF set by PsxRates[i] = (int)std::min(rate, (s64)0x3FFFFFFFLL);. So we end up with 0x40000000 Value.

Since nothing else seems to happen there. My guess is that PsxRates values should be allowed up to 0x40000000 instead of 0x3FFFFFFF. I tested this, and patch is not longer required to go in-game. Although this still don't fix delay which is most likely EE or IOP timing issue. 

This seems to be correct, because there is no other way to end up with 0x3FFFFFFF while leaving case 2. But i don't feel that I want to PR this. Too many things can break if i'm wrong. :P

**Comment by refractionpcsx2** [comment]
values on the SPU are generally -0x4000 to 0x3fff, so I'm not sure 0x4000 makes much sense :P

**Comment by Goatman13** [comment]
Then i don't know how real PS2 end up with 0x3FFF "ADSR Value". Maybe 0x4000 should be allowed there for situations where it is used to decrease volume? So its like doing Value + -0x4000 or something.

**Comment by refractionpcsx2** [comment]
Yeah I dunno, not even sure where this rate table came from, maybe a hangover from PCSX :P  ADSR tables and handling are not my forté, but yeah I'm not sure what the values should be personally...

**Comment by Ziemas** [comment]
I think it might indeed be off by one, if the decay rate is 0 I think the decrease should be 0x4000 per step. so `0x7fff (ADSR max) - 0x4000` is 0x3fff.

With the decay target set to 0x6800 `((0xc + 1) << 11)` anything below that would cause the ADSR to switch phase, so they're being very silly by expecting a value lower than that.

Note that unlike pcsx2 the spu uses 16bit math for ADSR, and the rate table was from before the algorithm was figured out. (although the table was at least improved since the pcsx days)

**Comment by refractionpcsx2** [comment]
maybe this line in the Init function is the culprit? :D
`PsxRates[i] = (int)std::min(rate, (s64)0x3fffffffLL);`

**Comment by Buzzardsoul** [comment]
> So I guess this line in the Init function is the culprit? `PsxRates[i] = (int)std::min(rate, (s64)0x3fffffffLL);`

Changing it to `PsxRates[i] = (int)std::min(rate, (s64)0x40000000LL);` has also managed to start the game without the need for the startup fix on my end too. I wonder if it truly is the real behavior of the original hardware.

**Comment by refractionpcsx2** [comment]
the startup fix? You mean the patch? An audio volume caused a TLB miss? O_o

**Comment by Buzzardsoul** [comment]
> the startup fix? You mean the patch? An audio volume caused a TLB miss? O_o

Seems like it did lmao.

**Comment by refractionpcsx2** [comment]
I've seen some strange things in my time, that one is top 10, at least!

**Comment by RedDevilus** [comment]
Spyro can be quite sensitive, like when I tried skipping a FMV it hardcrashed. No other games seem to do that.

**Comment by refractionpcsx2** [comment]
I tell a lie, the EU version is just broken, the US version does indeed work without a patch

**Comment by Buzzardsoul** [comment]
Sorry for flooding this issue, I've tested it on both a debug build and a release build and it's working on both without the patch being present in the gameindex file and enable compatibility patches being unchecked

**Comment by refractionpcsx2** [comment]
I mean, it does this on the EU version

![image](https://user-images.githubusercontent.com/6278726/231894228-d033154d-c162-42cb-b427-b52e1e7f630c.png)

but it only happens on the EU one, you *can* continue, if you ignore them :P

**Comment by Buzzardsoul** [comment]
Likely another issue that's not related to the black screen.

**Comment by refractionpcsx2** [comment]
yeah it is unrelated

**Comment by Buzzardsoul** [comment]
In regards to the TLB miss, The executable offset of 0x002C7D24 Is setting the v0 register's value to 0x0 while on real hardware it is supposed to be 0x00309DF0

**Comment by refractionpcsx2** [comment]
yeah it hits that code a few times, I haven't traced back the bad value to where it comes from.

**Comment by Buzzardsoul** [comment]
> yeah it hits that TLB miss code a few times, I haven't traced back the bad value to where it comes from.
> 
> (I know you deleted your comment, but I'm still putting it here :P)

Because it was very rushed lolz. Anyways, initial results show that 0x002c7e24 probably never gets hit on real hardware :octocat: 
I'll update this comment if I find anything new

**Comment by Goatman13** [comment]
Only PAL version TLB miss because problem is in CSpyroWorld::UpdateMenuLanguageSelect(float), NTSC skip that menu.

Right before that game try to read 0 sectors. 
```
cdvdWrite04: NCMD CdReadDVDV (8) (ParamP = b)
DvdRead > startSector=536681, seekTo=535510 nSectors=0, RetryCnt=100, Speed=2x(CLV), ReadMode=3(0) SpindleCtrl=2
Bad Sector Count Error
```

**Comment by refractionpcsx2** [comment]
it does, yep :P I'm not sure if it just does that, or if that's an error and it should be sending a sector count.

**Comment by Buzzardsoul** [comment]
@Goatman13 I have messed around with the SIF and the IOP quite extensively, I can only guess that the cutscene audio not being in sync is the fault of the SPU2. I mean, think about it, the game having an ADSR check is there for a reason. And I'll conclude with the fact that I just can't wrap my head around where the issue lies in the SPU2's code.

**Comment by Goatman13** [comment]
> @Goatman13 I have messed around with the SIF and the IOP quite extensively, I can only guess that the cutscene audio not being in sync is the fault of the SPU2. I mean, think about it, the game having an ADSR check is there for a reason. So, I'll conclude with the fact that I just can't wrap my head around where the issue lies in the SPU2's code but someone should probably check that part of the source code for inaccuracies.
> 
> Edit: and yes I'm aware that this is not relevant to the TLB miss and thank you for taking the time to check it out.

Game launch dialog/sound independently, at the same time EE is supposed to get flag that "cinematic is ready" by that ADSR 0x3fff stuff. Then EE side trigger rendering. You can test it on pcsx2 version without ADSR fix, if you disable patch sound play anyway with black screen. Game just expect EE to get there when it expect to be there. That should be better with +3 cyclerate, but is not. Maybe because its not tight loop to catch it up soon enough. Idk, from SPU2 side it looks correct now. 


> it does, yep :P I'm not sure if it just does that, or if that's an error and it should be sending a sector count.

Game request 0 sectors from cw_opt0.irx (custom streaming module, total surprise :P ). Seems to be intentional, sceCdRead return 0x6 , and module is happy that is not 0. That where error detection seems to end... Module have cdvd error detection itself expecting CDVD_STATUS_EMERGENCY to be set before going into any error check, setting that doesn't change anything for me. Module itself when in "emergency" perform check only against:

0x14: SCECdErCUD - Command is not valid for current disc
0x30: SCECdErREAD - Error while reading
0x31: SCECdErTRMOPN - Tray has been opened
0xFD: SCECdErREADCF - ?

So really no idea about that one.

Attached save state (slot3) is at point when previous read command is almost sent, next command is problematic one. Called from 0008E4B8 on IOP which = 0x00001888 of cw_opt0.irx, 00027368 is where sceCdRead is at iop dump.
[SLES-51043 (0EF1D4BA).03.zip](https://github.com/PCSX2/pcsx2/files/11240377/SLES-51043.0EF1D4BA.03.zip)

**Comment by Buzzardsoul** [comment]
Seems that this issue is no longer.

**Comment by refractionpcsx2** [comment]
Good to know, the TLB miss is still a problem though

**Comment by Goatman13** [comment]
PAL version is still have sound desync for me, even on press start screen screen. Tested on v1.7.4651 without using savestates.

**Comment by refractionpcsx2** [comment]
Just to note the sound sync thing is still a thing, though the TLB misses are completely gone these days on all versions, the bad sector count error is still there but it seems to be non-fatal, it may be that it's suppose to unintentionally be there.

I tried messing around with EE counters/timers, cycle rates etc to try and work out what causes the video to get ahead of the audio, but this was to no avail.

I keep wondering if there's some PAL thing we're missing and still doing NTSC-like.  I know it gets affected by HBLANK timing and EE cycle rates etc, but Valkyrie Profile 2 also has the video blasting ahead of the audio (I also achieved it by doing chunky cycle delays on the IPU, but it felt like too much and the "okay" area was a very small amount of wiggle room, I don't believe the game is this sensitive to IPU sync).  I'm not sure what it could be that we're not taking in to account, though, if that is the case.

Another case of this is Time Crisis 2, which does seem to be affected by EE cycle rate, again could be a red herring, but the audio gets ahead of the 3D rendering in this, rather than the other way around, then it starts making noises like the SPU overran, but the SPU DMA timings have changed a fair bit recently and it's actually no better or worse than it used to be, so again, another mystery.

Is there some other clock rate we're not dealing with, or something else reliant on vblank/hblank speeds but we're doing it based on CPU cycles some how? I really don't know..

Anyway, sorry for the ramblings, just needed to get this down somewhere.

**Comment by Goatman13** [comment]
Not sure if i remember correctly, but I think COP0 performance counter affects sound related timings in Spyro. This can be wrong i don't have access to game now. I mean one of those games you mentioned for sure is affected by that, but i think it was Spyro.

**Comment by refractionpcsx2** [comment]
I think I tried dividing those by two :D maybe I didn't check that game, I don't remember!

**Comment by refractionpcsx2** [comment]
I just tried it, made no difference :(

**Comment by Buzzardsoul** [comment]
I thought maybe finding which pr fixed the desync for the US version could be useful for finding a fix for EU version, and it seems that #8657 is the one, because v1.7.4381's sound is perfectly in sync as opposed to v1.7.4380's.

**Comment by refractionpcsx2** [comment]
A GS dump isn't going to show audio sync.

Anyway, the SLES version is still massively wrong.

**Comment by Loy2up** [comment]
I've tested the EU version on OPL, and seemingly, the audio/video desync is there too. I'm only using mode 2 (Synchronous mode) with it. Hope that helps you find the issue.

**Comment by CRASHARKI** [comment]
I found a way to fix the sped up cutscenes. While researching the game files, I noticed that the European version has this in the cutscene and title chunk files:
```
# For PAL only
TimeScaleFactor:	0.833
```
So, using HxD, I replaced:
"TimeScaleFactor:	0.833"
with:
"#imeScaleFactor:	0.833"
directly in the ISO, and voilà! The animations are no longer sped up, and the audio is perfectly in sync.
Unfortunately, I don't know how this could be implemented in the emulator.

**Comment by Goatman13** [comment]
Interesting find. 30 x 0.833 is actually 24.99. It's like game expect some counter to work in "NTSC" mode and translates it by itself to "PAL" mode. Weird stuff, because issue with counters like that would probably affect many more games. 

Edit: That really doesn't make sense (but it works, confirmed already). This value is used as a divisor for value from performance counter. PCCR is set to count processor cycles on pcr0, in user/kernel/exc lv1 modes. PCR1 is disabled (set to No Event).

This is kinda what i mentioned here when i tried to blindly debug it: https://github.com/PCSX2/pcsx2/issues/8286#issuecomment-1802686508

All this doesn't really make much sense because they count cpu cycles that shouldn't vary between PAL/NTSC.
Only time when value from pcr0 is affected is when time converted to float is higher than 0.066666, which on pcsx2 is never true (it is around 0.02), then time is forced to be 0.066666. I tried overclocking etc, but it doesn't really change much, value is still not higher than 0.02xxxx.

Since i really have no idea what's going on, i decided to write a patch to force TimeScaleFactor to 1.0. I'm not sure if other part of game will be happy with it, someone more familiar with this game should test it.

```
// Force TimeScaleFactor to 1.0 - CRC 0EF1D4BA
patch=1,EE,001E76F4,word,46001006
```
 
One more edit: I think that setting this to 1.0 is a hack but should be ok for now. Real issue is probably on other side, on IOP or SPU2. Since it's impossible that cycle counter will vary per framerate.

**Comment by Ziemas** [comment]
Looks like it calibrates the cycles per vblank value on its own.

```c
// 0x193640
void c6Time::Create(int param_1)
{
    int iVar1;
    
    if (_offset == NULL) {
        _nChannels = param_1;
        _offset = (uint *)operator_new[](param_1 << 2);
        for (iVar1 = 0; iVar1 < _nChannels; iVar1 += 1) {
            _offset[iVar1] = 0;
        }
        sceGsSyncV(0);
        scePcStart(0x8008003e);
        sceGsSyncV(0);
        _cyclesPerVblank = PCR0;
        c6CLOCKS_RATIO = 1.0 / ((float)PCR0 * 59.9);
        scePcStop();
    }
    return;
}
```
_cyclesPerVblank = 0x4AFFE9
RATIO = 0x31696830 (~0.00000000339651874552)

This seems wrong for PAL though? It calculates this while PCSX2 is set to 60 VPS, is this an issue?

**Comment by Ziemas** [comment]
Disabling fast boot and booting with EU BIOS I get _cyclesPerVblank = 0x59FFE9 instead.

And the timing also seems much better.

So it seems that the game expect that OSDSYS or whatever has already set the system to a PAL video mode.

**Comment by Goatman13** [comment]
Nice work. I got so distracted by that second spot where it measures time during cutscene playback that I didn’t even look into the other one.
